src/main/java/com/google/cloud/tools/maven/DeployIndexMojo.java

```diff
@@ -16,25 +16,22 @@
 
 package com.google.cloud.tools.maven;
 
-import com.google.cloud.tools.appengine.api.AppEngineException;
-import com.google.cloud.tools.appengine.api.deploy.AppEngineDeployment;
-import com.google.cloud.tools.appengine.api.deploy.DeployProjectConfigurationConfiguration;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugins.annotations.Execute;
 import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 
 /** Stage and deploy index.yaml to Google App Engine standard or flexible environment. */
 @Mojo(name = "deployIndex")
 @Execute(phase = LifecyclePhase.PACKAGE)
-public class DeployIndexMojo extends AbstractSingleYamlDeployMojo {
+public class DeployIndexMojo extends AbstractDeployMojo {
 
-  protected void doDeploy(
-      AppEngineDeployment appEngineDeployment,
-      DeployProjectConfigurationConfiguration configuration) {
-    try {
-      appEngineDeployment.deployIndex(this);
-    } catch (AppEngineException ex) {
-      throw new RuntimeException(ex);
-    }
+  @Override
+  public void execute() throws MojoExecutionException, MojoFailureException {
+    AppEngineDeployer deployer = AppEngineDeployer.Factory.newDeployer(this);
+    deployer.configureAppEngineDirectory();
```

Comment:
So we don't actually need to stage on flex executions of this task do we?

I wonder if `stage` should just be an internal call of `deployer.deployIndex`, similarly `configureAppEngineDirectory` can be internal to deployer?

Reply:
The reason I left `configureAppEngineDirectory` separate is because `deploy` and `deployAll` don't call it before `stage` (they use src/main/appengine regardless of standard or flexible).

I also originally kept `stage` by itself because testing was difficult (it kept deleting my test files for the `deployAll` test when it cleared the staging directory). I could split things out into more methods to make testing deploy functionality without staging easier, but then I'd end up with double the number of deployment methods in these classes, which could be messy.

Reply:
Oh right... you need to call it before stage.

Could we do something like this?
```
class Stager {}
class Deployer {
  Stager stager
}
```

Reply:
Ah yeah, that's a good idea. Solves the problem from your other comment as well.